madprops/serializers.py

```diff
@@ -61,95 +91,160 @@ class Property(model):
+            'name': prop_name
         }
-        if not model.objects.filter(**filters).update(value=obj.value):
-            obj.save(**kwargs)
-
 
-class NestedPropertySerializer(PropertySerializer):
-    """Version of PropertySerializer for nested resources
-
-    Intended to be used as a base class for serializer for property resource
-    exposed as a nested resource.
-    """
+        # If it already exists - update it's value. Otherwise - create a new
+        # property.
+        prop = self.Meta.model.objects.filter(**filters).first()
+
+        if prop_name in self.opts.read_only_props:
+            return prop
+
+        if prop:
+            prop.value = property_data['value']
+            prop.save()
+        else:
+            prop = self.Meta.model.objects.create(**property_data)
+
+        return prop
+
+    def to_internal_value(self, data):
+        # Data can be in two formats here depending on many=True/False:
+        #   - {<prop_name>: <prop_value>}
+        #   - {'name': <prop_name>, 'value': <prop_value>}
+        # Convert to the format accepted by standard DRF serializers.
+        if not('name' in data and 'value' in data):
+            data = self._to_extended_dict(data)
+
+        # Handle JSON fields (value encoded as JSON).
+        if data['name'] in self.opts.json_props:
+            data['value'] = json.dumps(data['value'])
+
+        data = self._add_parent_obj_field(data)
+        return super(PropertySerializer, self).to_internal_value(data)
+
+    def _to_extended_dict(self, data):
+        """ Convert dictionary of properties:
+        {<prop_name>: <prop_value>} ->
+            {'name': <prop_name>, 'value': <prop_value>}
+        """
+        prop_name, prop_value = data.items()[0]
+        return {'name': prop_name, 'value': prop_value}
+
+    def _add_parent_obj_field(self, data):
+        # Property requires ID of parent object (properties owner, e.g. User
+        # for user's properties case). Take it from context, which is passed
+        # from view.
+        parent_obj_field = self.opts.parent_obj_field
+        parent_id_field = parent_obj_field + '_id'
+        data[parent_obj_field] = self.context.get(
+            'parent_id') or self.context['view'].kwargs.get(
+                parent_id_field)
+        return data
 
-    _options_class = NestedPropertySerializerOptions
 
-    def __init__(self, **kwargs):
-        super(NestedPropertySerializer, self).__init__(**kwargs)
-        self.many = True
+class PropertiesOwnerSerializer(ModelSerializer):
+    """Base class for "parent" serializers.
 
-    def _from_native_hook(self, data):
-        return data
+    If you want to use PropertySerializer with a parent model (e.g. as a
+    nested serializer for User serializer), and have this field writable  -
+    you have to inherit you parent serializer from this class, not
+    ModelSerializer.
+    """
+    def create(self, validated_data):
```

Comment:
I don't think we should copy-and-paste entire `create()`, let's override `raise_errors_on_nested_writes()`, call parent `create()` and after that `_save_properties`().